src/Psalm/Internal/Analyzer/Statements/Expression/Fetch/VariableFetchAnalyzer.php

```diff
@@ -531,26 +540,231 @@ public static function getGlobalType(string $var_id): Union
+                'REQUEST_URI'          => $non_empty_string_helper,
+                'PHP_AUTH_DIGEST'      => $non_empty_string_helper,
+                'PHP_AUTH_USER'        => $non_empty_string_helper,
+                'PHP_AUTH_PW'          => $non_empty_string_helper,
+                'AUTH_TYPE'            => $non_empty_string_helper,
+                'PATH_INFO'            => $non_empty_string_helper,
+                'ORIG_PATH_INFO'       => $non_empty_string_helper,
+                // misc from RFC not included above already http://www.faqs.org/rfcs/rfc3875.html
+                'CONTENT_LENGTH'       => $string_helper,
+                'CONTENT_TYPE'         => $string_helper,
+                // common, misc stuff
+                'FCGI_ROLE'            => $non_empty_string_helper,
+                'HOME'                 => $non_empty_string_helper,
+                'HTTP_CACHE_CONTROL'   => $non_empty_string_helper,
+                'HTTP_COOKIE'          => $non_empty_string_helper,
+                'HTTP_PRIORITY'        => $non_empty_string_helper,
+                'PATH'                 => $non_empty_string_helper,
+                'REDIRECT_STATUS'      => $non_empty_string_helper,
+                'REQUEST_SCHEME'       => $non_empty_string_helper,
+                'USER'                 => $non_empty_string_helper,
+                // common, misc headers
+                'HTTP_UPGRADE_INSECURE_REQUESTS' => $non_empty_string_helper,
+                'HTTP_X_FORWARDED_PROTO'         => $non_empty_string_helper,
+                'HTTP_CLIENT_IP'                 => $non_empty_string_helper,
+                'HTTP_X_REAL_IP'                 => $non_empty_string_helper,
+                'HTTP_X_FORWARDED_FOR'           => $non_empty_string_helper,
+                'HTTP_CF_CONNECTING_IP'          => $non_empty_string_helper,
+                'HTTP_CF_IPCOUNTRY'              => $non_empty_string_helper,
+                'HTTP_CF_VISITOR'                => $non_empty_string_helper,
+                'HTTP_CDN_LOOP'                  => $non_empty_string_helper,
+                // common, misc browser headers
+                'HTTP_DNT'                => $non_empty_string_helper,
+                'HTTP_SEC_FETCH_DEST'     => $non_empty_string_helper,
+                'HTTP_SEC_FETCH_USER'     => $non_empty_string_helper,
+                'HTTP_SEC_FETCH_MODE'     => $non_empty_string_helper,
+                'HTTP_SEC_FETCH_SITE'     => $non_empty_string_helper,
+                'HTTP_SEC_CH_UA_PLATFORM' => $non_empty_string_helper,
+                'HTTP_SEC_CH_UA_MOBILE'   => $non_empty_string_helper,
+                'HTTP_SEC_CH_UA'          => $non_empty_string_helper,
             ]);
+
+            // generic case for all other elements
+            $detailed_type->previous_key_type = Type::getNonEmptyString();
+            $detailed_type->previous_value_type = Type::getString();
```

Comment:
Are we sure it's always Strings? You documented a float above. Mixed could be more appropriate?

Reply:
Maybe `scalar`? I haven't looked into how PHP populates `$_SERVER`, but I would _guess_ they would always be scalars.

Reply:
note that scalar excludes null, so possibly scalar|null

Reply:
Many of these come from apache2/nginx (some of them from those servers themselves, some forwarded from the client), is it possible for those to set integer/float/whatever else values or is it forced to be a string?

Reply:
We'll fix that if we get an example then, thanks for checking all that!